stingray/crossspectrum.py

```diff
@@ -348,26 +398,63 @@ def _make_crossspectrum(self, lc1, lc2):
                 nphots1_all.append(nphots1_sep)
                 nphots2_all.append(nphots2_sep)
 
-            cs_all = np.hstack(cs_all)
+            self.cs_all = np.hstack(self.cs_all)
             nphots1_all = np.hstack(nphots1_all)
             nphots2_all = np.hstack(nphots2_all)
 
 
-        m = len(cs_all)
+        m = len(self.cs_all)
         nphots1 = np.mean(nphots1_all)
         nphots2 = np.mean(nphots2_all)
 
-        cs_avg = np.zeros_like(cs_all[0].cs)
-        for cs in cs_all:
+        cs_avg = np.zeros_like(self.cs_all[0].cs)
+        for cs in self.cs_all:
             cs_avg += cs.cs
 
         cs_avg /= np.float(m)
 
-        self.freq = cs_all[0].freq
+        self.freq = self.cs_all[0].freq
         self.cs = cs_avg
         self.m = m
-        self.df = cs_all[0].df
-        self.n = cs_all[0].n
+        self.df = self.cs_all[0].df
+        self.n = self.cs_all[0].n
         self.nphots1 = nphots1
         self.nphots2 = nphots2
 
+    def coherence(self):
+        """
+        Compute an averaged Coherence function of cross spectrum by computing
+        coherence function of each segment and averaging them. The return type
+        is a tuple with first element as the coherence function and the second
+        element as the corresponding uncertainty[1] associated with it.
+
+        Note : The uncertainty in coherence function is strictly valid for
+               Gaussian statistics only.
+
+        Returns
+        -------
+        tuple : tuple of np.ndarray
+            Tuple of coherence function and uncertainty.
+
+        References
+        ----------
+        .. [1] http://iopscience.iop.org/article/10.1086/310430/pdf
+
+        """
+        if self.m < 50:
+            utils.simon("Number of segments used in averaging is "
+                        "significantly low. The result might not follow the "
+                        "expected statistical distributions.")
+
+        # Calculate average coherence
+        coh = np.zeros_like(self.cs_all[0].coherence())
+        for acs in self.cs_all:
+            coh += acs.coherence()
```

Comment:
I think this definition isn't quite correct. Looking at Equation 7 in [Nowak 1999](http://iopscience.iop.org/article/10.1086/306610/pdf), I think the coherence for a large number of segments should be the square of the absolute value of the average of `cs.unnorm_cross` divided by the average of the power spectra of `lc1` times the average of the power spectra of `lc2`. So basically, compute the average square of `cs.unnorm_cross` for all segments, then compute `AveragedPowerspectrum` for each of the light curves, and then compute the coherence from there. Does that make sense? It might look similar, but I think in practice it can lead to subtle errors to not do it that way.